driver/failover_handler.cc

```diff
@@ -537,14 +537,30 @@ bool FAILOVER_HANDLER::failover_to_writer(const char*& new_error_code, const cha
         error_msg = "The active SQL connection was lost.";
         return false;
     }
+
+    const auto new_topology = result->new_topology;
+    const auto new_host = new_topology->get_writer();
+
     if (result->is_new_host) {
         // connected to a new writer host; take it over
-        current_topology = result->new_topology;
-        current_host = current_topology->get_writer();
+        current_topology = new_topology;
+        current_host = new_host;
+    }
+    const auto filtered_topology = this->topology_service->get_filtered_topology(new_topology);
+    const auto allowed_hosts = filtered_topology->get_instances();
+    if (std::find(allowed_hosts.begin(), allowed_hosts.end(), new_host) == allowed_hosts.end()) {
+        new_error_code = "08S01";
```

Comment:
A comment here may be helpful to explain why this particular error code string is being used.